Dev with asterix optional

kstppd opened the pull request:
# Asterix Compression

Asterix provides lossy compression for VDFs. Current compression methods:
- `ZFP`
- `MLP_MULTI`
- `OCTREE`

TODO:   
- `HERMITE` which is pretty much ready but maybe will be added in a future PR.


**All methods work for both CPU and GPU Vlasiator. However running the MLP on CPUs is strongly discouraged for performance reasons**

The install libraries scripts will install ZFP and OCTREE automagically but the MLP has to be installed manually. It requires either `nvcc` or `hipcc` to be installed on the system. For instructions look at the [Asterix](https://github.com/fmihpc/asterix) repo.


```bash
pip install meson
git clone https://github.com/fmihpc/asterix
cd asterix/  
mkdir subprojects
meson wrap install gtest
meson wrap install curl
meson wrap install spdlog 
meson wrap install fmt
meson setup build  --buildtype=release -Dtiny_cuda_arch_code=70/80/90
meson compile -C build --jobs=8
meson test -C build
```
Some one-liner that is good for NVIDIA systems and saves you from the build system havoc:    
```
cd asterix/
nvcc src/vdf_compressor_nn.cu -DUSE_GPU -DSKIP_HOSTBLAS  -DTINYAI_MEMORY_GB=64  -DNOPROFILE --std=c++20 -I./include -arch=sm_86  --shared  -Xcompiler=-fPIC -o libvlasiator_vdf_compressor_nn.so -lcublas 
```

The compression method is selected at runtime with `--Asterix.method`.

## Compile time Flags
Compression methods are optional and enabled at compile time.

```bash
-DASTERIX_ZFP
-DASTERIX_OCTREE
-DASTERIX_MLP
```

Example: For an example look into MAKE/Makefile.kstppd

## Compression Parameters

```text
--Asterix.mlp_layers
    Hidden layer architecture for the MLP.
    Example: 32,32,16

--Asterix.tol
    Tolerance used by the MLP compressor.

--Asterix.octree_tolerance
    Tolerance used by OCTREE compressor.

--Asterix.max_epochs
    Maximum MLP training epochs to be used.

--Asterix.fourier_order
    Fourier feature order used by the MLP compressor.

--Asterix.interval
    Compression interval in seconds [Deprecated and unused].

--Asterix.state
    Enable or disable compression. 
    0 = disabled
    1 = enabled

--Asterix.method
    Compression method to use.
    Example: ZFP, MLP_MULTI, OCTREE

--Asterix.max_vdfs_per_nn
    Maximum number of VDFs grouped into one neural network
    in multi-regression mode.
```

## I/O Compression Options
These options enable compression for different output file types.

```text
--io.system_write_distribution_compressed 
    Compress VDFs written to bulk distribution files.

--io.system_write_restart_compressed
    Compress VDFs written to restart files.

--io.system_write_recovery_compressed
    Compress VDFs written to recovery files.
```

kstppd committed:
Modify Makefile for local machine

kstppd committed:
Ignore compile commands

kstppd committed:
update Makefile

kstppd committed:
update ignore list

kstppd committed:
update ignore list again

kstppd committed:
Asterix runtime compression using fourier MLP

kstppd committed:
Asterix runtime compression using ZFP

kstppd committed:
Add method to probe MLP size. To be used with transfer learning

kstppd committed:
Create a feedback mechanism for weights to be passed aroung so we can do some transfer learning

kstppd committed:
Use master rank instead of 0

kstppd committed:
Allow for a more generic driver from vdf compression. Select compression scheme in cfg

kstppd committed:
Actually get number of SC. AMR later

kstppd committed:
Enable transfer learning

kstppd committed:
More reasonable zfp tol

kstppd committed:
Add forgotten block adjustment back in

kstppd committed:
Some fixes and better error reporting

kstppd committed:
Lumi Makefiles

kstppd committed:
Clang pass

kstppd committed:
Bump to c++20

kstppd committed:
Some updates and better vcoords handling in the interface

kstppd committed:
interface for exposing vdfs to compression methods using dense ordered meshes

kstppd committed:
Add code for using different resolution in training vs inference

kstppd committed:
Add code for using different resolution in training vs inference

kstppd committed:
Yup. Do not overwrite harmonics

kstppd committed:
Report timings in logfile with chrono

kstppd committed:
First clean up pass after Lumi hackathon in Brussels

kstppd committed:
more cleanup

kstppd committed:
Update macros and some vlasiator changes for LUMI HIP

kstppd committed:
remove device selection -- will be done automagically

kstppd committed:
Add functionality to use one MLP per multiple VDFs.

kstppd committed:
Please, do not hash floating point numbers...

kstppd committed:
Update lumi hip file

kstppd committed:
LocalID to the rescue

kstppd committed:
Fix compression ratio calculation

kstppd committed:
Fix compression ratio calculation

kstppd committed:
Clean up

kstppd committed:
Clean up and revert to using block GID

kstppd committed:
Formatting pass

kstppd committed:
Clean up again

kstppd committed:
some vdf norm settings

kstppd committed:
Downsampling for multi vdfs

kstppd committed:
Formatting pass

kstppd committed:
Buggy union fixed

kstppd committed:
More cleanups

kstppd committed:
New structs to simplify interface

kstppd committed:
Restructuring to expose separate API for compression libs

kstppd committed:
Octree interface

kstppd committed:
Reconctruct VDF from dense mesh

kstppd committed:
Disable zoom in dense extraction for now since the reconstruction cannot yet handle that

kstppd committed:
Remove debug print

juhanikataja committed:
link to octree compressor code

kstppd committed:
Merge pull request #5 from juhanikataja/asterix_hack_3

link to octree compressor code

kstppd committed:
Makefile update for octree

kstppd committed:
Fix bug in union initialization

kstppd committed:
some small testcases with higher harmonics

kstppd committed:
more fixes

kstppd committed:
more fixes 2

juhanikataja committed:
ASTERIX: some tollerance and tolerrance typo fixes and parameter renamings

kstppd committed:
updates to normalization routine

kstppd committed:
Reduce some info across ranks for logging

kstppd committed:
Dump debug methods

kstppd committed:
Defrag the pool before inference

kstppd committed:
Merge remote-tracking branch 'juhani/asterix_hack_3' into tmp

kstppd committed:
Merge branch 'compressiator_debug_dump_vdfs' into asterix_hack_3

kstppd committed:
Calculate shannon entropy

kstppd committed:
Pich MLP arch based on Shannon Entropy and a fudge parameter

kstppd committed:
temp

kstppd committed:
Allow multi regression on subspans

kstppd committed:
Ignore updates

kstppd committed:
Make copy ctors non explicit

kstppd committed:
Update fourier mappings and some more fiddling around

kstppd committed:
Add SIREN impl

kstppd committed:
New struct for VDFUnion(s)

kstppd committed:
Some updates to initialization

kstppd committed:
calc non maxwellianity

kstppd committed:
Cluster based on  non maxwellianity

kstppd committed:
Functionality to  center VDFs subtracting bulk velocity from vcoords

kstppd committed:
Merge dev

kstppd committed:
uddate submodules and local makefile

kstppd committed:
Clean up network size probing

kstppd committed:
Update Siren act function usage

kstppd committed:
Typos and std abs in the affine trasnforms calc.s

kstppd committed:
Once again the hell of supporting both cuda and hip namespaces

kstppd committed:
Revert "Update Siren act function usage"

This reverts commit 1e271b9c0acbb3aff5cbc252b9881a3dc7ddda20.

kstppd committed:
Another clean up pass with some modification to batch size and siren activations

juhanikataja committed:
use renamed T-Octree

juhanikataja committed:
update toctree interface

juhanikataja committed:
update toctree interface to include maxiter

kstppd committed:
Remove rust implementation

kstppd committed:
Clean up lib.cu

kstppd committed:
Remove single VDF compression since that is now covered by the _multi versions

kstppd committed:
Clean up

kstppd committed:
Return int instead of float for the bytes used by the MLP

kstppd committed:
Merge remote-tracking branch 'juhani/compressiator' into rename_octree

kstppd committed:
Update local makefile for updated octree

kstppd committed:
Update ignore list

kstppd committed:
Clean up compression ratio reduction over MPI

kstppd committed:
Better default tol for octree

kstppd committed:
Initial infra for IO

kstppd committed:
Move scaling and normalizing in vlasiator

kstppd committed:
Split compression and decompression for MLP in 2 separate steps. Remove
EGC guided comnpression

kstppd committed:
Split compression and decompression for MLP in 2 separate steps

kstppd committed:
Merge branch 'compressiator_io' of github.com:kstppd/vlasiator into compressiator_io

kstppd committed:
Remove some defs

kstppd committed:
Serialize VDFUnions (almost done)

kstppd committed:
Bug fix in index calculation. Thanks Juhani

kstppd committed:
Minor update

juhanikataja committed:
fix indexing in compression.cpp and treat toctree_compressor status

kstppd committed:
Disable reconstruction. Add ZFP and OCTREE to restart files only

kstppd committed:
Add bytes per cell to the restart file

kstppd committed:
Compressed state is per pop

kstppd committed:
WIP read in ZFP state from restart crashes.

kstppd committed:
ZFP IO done.

kstppd committed:
OCTREE proper metadata write

kstppd committed:
OCTREE io done

kstppd committed:
MLP IO done

kstppd committed:
Update Makefile for lumi

kstppd committed:
revert to less compression for mlp

kstppd committed:
IO for clustered MLPs

kstppd committed:
Proper error reporting for multi regression

kstppd committed:
Better error calculation in tinyAI

kstppd committed:
even better tinyai

kstppd committed:
even better tinyai 2

kstppd committed:
Interim commit. Clustered mlp read IO is broke

kstppd committed:
Add elu

kstppd committed:
update

kstppd committed:
const correctness and better memory footprint

kstppd committed:
Clustered MLP IO done

kstppd committed:
Faster shuffling in tinyAI

kstppd committed:
Faster shuffling in tinyAI

kstppd committed:
Remove driver types header

kstppd committed:
forgotten return

kstppd committed:
Ooopsie

kstppd committed:
Fix error calc mask mistake. Shuffler buggy  for small batchizes

kstppd committed:
tmp fix

kstppd committed:
Fix data shuffler damn it

kstppd committed:
updates on lumi

kstppd committed:
fix conflict

kstppd committed:
do not generate blocks that are not needed  + patience in lic.xu

kstppd committed:
remove redundant code

kstppd committed:
store block complement for octree

kstppd committed:
store block complement for octree 2

kstppd committed:
stable tsi runs on lumi

kstppd committed:
update eigen

kstppd committed:
add miniapp for multi regression against pytorch

kstppd committed:
update tinyai

kstppd committed:
updates in tinyAI and mini app

kstppd committed:
mini app update

kstppd committed:
dynamic dispatch based on activation function part 1

kstppd committed:
dynamic dispatch based on activation function part 2

kstppd committed:
Allow batched inferense because OOM was real

kstppd committed:
Merge branch 'compressiator_2' of github.com:kstppd/vlasiator into compressiator_2

kstppd committed:
Revert "mini app update"

This reverts commit 13b169023e531fbf2900af127472ecac7938d005.

kstppd committed:
Revert "dynamic dispatch based on activation function part 2"

This reverts commit d1a4868ce3ed0151520e8c395cc97666df8d37f4.

kstppd committed:
Revert "dynamic dispatch based on activation function part 1"

This reverts commit 976d30ab500312421b797d5d3a295de83aaaa123.

kstppd committed:
Update toctree call signature

kstppd committed:
Fix previous commit

kstppd committed:
Merge branch 'compressiator_2' of github.com:kstppd/vlasiator into compressiator_2

kstppd committed:
lumi merged adamw

kstppd committed:
lumi merged forward

kstppd committed:
Merge branch 'compressiator_2' of github.com:kstppd/vlasiator into compressiator_2

kstppd committed:
formatter pass

kstppd committed:
updates to miniapp

kstppd committed:
WIP1

kstppd committed:
set up for threads

kstppd committed:
Now we thread over clusters. Tests pending

kstppd committed:
Formatter pass

kstppd committed:
Formatter pass 2

kstppd committed:
Formatter pass 3

kstppd committed:
Addd log-cosh loss function

kstppd committed:
Final cleanup

kstppd committed:
Add mnist unit test as mini app

kstppd committed:
update unit

kstppd committed:
reduce mem footprint of linear layers

kstppd committed:
Make layers polymorphic. Unfortunately could not be avoided

kstppd committed:
Formatting plus warning handling

kstppd committed:
Formatting plus warning reductions

kstppd committed:
fix mnist and a bug in = operator of matrix

kstppd committed:
remove tinyAI and compression lib from vlasiator

kstppd committed:
mempool stays

kstppd committed:
Revert "mempool stays"

This reverts commit 312eef1f06afbbd1769c3b319b9ebfa4e7cf0629.

kstppd committed:
update makefile

kstppd committed:
clean up pass wip 1

kstppd committed:
clean up pass wip 2, float training works

kstppd committed:
clean up pass wip 3, standardize

kstppd committed:
clean up pass wip 3, standardize

kstppd committed:
clean up pass wip 4

kstppd committed:
clean up pass wip 5

kstppd committed:
scale after collection

kstppd committed:
fmt

kstppd committed:
lumi fixes

kstppd committed:
lumi fixes

kstppd committed:
compute variance and sigma

kstppd committed:
Allow MLP to chuck VDFs similarly to MLP_MULTI

kstppd committed:
Skip do not compute cells in octree

kstppd committed:
fix omp loop limits

kstppd committed:
skip do not compute and fixed threading in MLP compression

kstppd committed:
Allow backwards compatibility with non Asterix restarts

kstppd committed:
threading fixes plus sorting of local cells

kstppd committed:
cleanup and remove some old commented out code

kstppd committed:
fix exclusive scan overflows

kstppd committed:
guard against bad MLP state reads

kstppd committed:
change normalization scheme

kstppd committed:
Merge pull request #13 from fmihpc/master

Roll the branch forward to current master

kstppd committed:
Clean up VDF norms in compression tools

kstppd committed:
Merge branch 'compressiator_threading_updates' of github.com:kstppd/vlasiator into compressiator_threading_updates

kstppd committed:
Add MLP specific DROs

kstppd committed:
reorder block writes to also iclude new DROs

kstppd committed:
add spell checker for typos in cfg

kstppd committed:
debugs from mpirank

Commit:
update phobos makefile

kstppd committed:
fix typo in error message. coding on a plane is difficult

kstppd committed:
merge dev into asterix

kstppd committed:
Revert "add spell checker for typos in cfg"

This reverts commit 967060e9bfcf3ab4d8cf80f6a98216d5875ad4bb.

kstppd committed:
Merge branch 'dev' of https://github.com/fmihpc/vlasiator into dev_with_asterix

kstppd committed:
Merge remote-tracking branch 'fmihpc/dev' into dev_with_asterix

kstppd committed:
Testing optional asterix features

kstppd committed:
indent Asterix enum

kstppd committed:
Enable bulk file vdfcompression

kstppd committed:
Add options for compressing bulk/recovery/restart

kstppd committed:
oopsie

kstppd committed:
udpate makefile too

kstppd committed:
Update local makefile

kstppd committed:
Merge branch 'dev' of https://github.com/fmihpc/vlasiator into dev_with_asterix_optional

kstppd committed:
Update CI docker image id

kstppd committed:
enable submodule checkout in CI

kstppd committed:
fix eigen cmake path

kstppd committed:
debug 1

kstppd committed:
debug 2

kstppd committed:
remove debug from build libraries

kstppd committed:
remove tinyAI mini app

ursg commented:
Can we get an update on the functionality here? Is this PR now in such a state that no specific Asterix features are enabled by default, so that we can "just merge it"? (Hmm, working CI would really help with this)

kstppd committed:
Merge branch 'dev' of https://github.com/fmihpc/vlasiator into dev_with_asterix_optional

kstppd committed:
Remove annoying debug printout of clustering scheme

kstppd committed:
Add fourier order and mlp arch to bulk files

kstppd committed:
update makefile for local machine

kstppd committed:
fix conflicts

kstppd committed:
remove ifdefs from Asterix enum because ofcourse they messed up the reading order

kstppd committed:
fix GPU compilation with Asterix shenanigans

kstppd committed:
Part2:fix GPU compilation with Asterix shenanigans

kstppd committed:
Handle linkage with new Asterix interface

kstppd committed:
fix iono mini app compilation with Asterix

kstppd committed:
update local cuda makefile

kstppd committed:
change enum order for asterix methods

kstppd committed:
resolve conflicts

kstppd committed:
fix conflict update to dev

kstppd committed:
Makefile changes. Disable clustering in MLP_MULTI temporarilly until I have a max number of allowed clusters

kstppd committed:
Makefile changes. Disable clustering in MLP_MULTI temporarilly until I have a max number of allowed clusters

petschge commented on vlasovsolver/vec.h:
```diff
@@ -220,7 +220,6 @@ typedef Vec16fb Vecb;
 #define VPREC 4
 #ifndef VEC_PER_PLANE
-const int VEC_PER_PLANE = (WID*WID/VECL);
 #endif
```
why remove that line?

kstppd replied:
I guess this was accidental. Reverting

petschge commented on spatial_cells/spatial_cell_gpu.hpp:
```diff
@@ -624,6 +628,7 @@ __global__ static void resize_and_empty_kernel (
 
       // Following functions adjust velocity blocks stored on the cell //
       void adjustSingleCellVelocityBlocks(const uint popID, bool doDeleteEmpty=false);
+      bool add_velocity_block(const vmesh::GlobalID& block,const uint popID);
```
Don't we use a lovely mix of snake_case and CamelCase?

kstppd replied:
Hmmm worth changing? We have this mix everywhere :)

petschge commented on submodules/fsgrid:
This is going to be so much fun to merge with the fsgrid changes...

kstppd replied:
This was accidental and has now been reverted! Thanks!

petschge commented on ioread.cpp:
```diff
@@ -354,22 +365,9 @@ bool checkScalarParameter(vlsv::ParallelReader& file,const string& name,T correc
    }
 }
 
-/** Read velocity block mesh data and distribution function data belonging to this process 
```
Why does that doc-comment get removed?

petschge commented on memory_report.cpp:
```diff
@@ -100,7 +100,8 @@ uint64_t get_node_free_memory(){
             // Skip to neww line
-            fscanf(in_file, "%*[^\n]\n");
+            auto _id=fscanf(in_file, "%*[^\n]\n");
+            (void)_id;
```
Is silencing the compiler warning really worth this? Maybe we should check the return value and act on it? And we might as well fix the typo in the comment while we're at it.

petschge commented on vlasiator.cpp:
```diff
@@ -1036,7 +1040,7 @@ int simulate(int argn,char* args[]) {
                   (uint)P::t,
                   true, // add the date of the file to the name
-                  P::restartStripeFactor) == false ) {
+                  P::restartStripeFactor,P::systemWriteRestartCompressed) == false ) {
```
one line per argument?

petschge commented on vlasiator.cpp:
```diff
@@ -1072,7 +1076,7 @@ int simulate(int argn,char* args[]) {
                   recoverCounter % P::recoverMaxFiles,
                   false, // overwrite so do not put date in file name
-                  P::restartStripeFactor) == false ) {
+                  P::restartStripeFactor,P::systemWriteRecoveryCompressed) == false ) {
```
one line per argument?

kstppd committed:
Revert fsgrid submodule update and handle some of the PR comments

kstppd committed:
Memory report update

kstppd committed:
Better asterix defaults

kstppd committed:
Better parameter descriptions

kstppd committed:
white space patch

kstppd committed:
one more makefile for me please

kstppd committed:
force sparsity in output if compression is on

kstppd committed:
update:force sparsity in output if compression is on

kstppd committed:
fix mempool include header

kstppd committed:
fix mempool include header

kstppd committed:
update some default asterix params

ursg committed:
Add apt-get dependency install on github_actions as well, for zfp.

Specifically, zfp and eigen refused to build without cmake, but I simply
copied the whole dependency line.

ursg committed:
Install dependencies *first*, then compile.

ursg committed:
No sudo on the ci container, use apt directly.

ursg committed:
Produce eigen cmake files before building zfp and octree.

ursg committed:
Actually checkout submodules for libraries_github

Since we now have dependencies that rely on eigen
